### PR TITLE
Switch (some) coding/encoding in conventions.py to use xarray.coding.

### LINKED ISSUE #1781: UnboundLocalError when opening netCDF file
#### Code Sample, a copy-pastable example if possible

```python
import xarray as xr
import netCDF4 as nc
product = '/Users/stefano/src/s5p/products/NO2/'\
          'S5P_OFFL_L2__NO2____20171107T195219_20171107T213349_00361_01_001107_20171108T122727.nc'

# opening the product with netCDF4 works fine
no2 = nc.Dataset(product)
no2.groups
# correctly shows groups' content
```
```
OrderedDict([('PRODUCT', <class 'netCDF4._netCDF4.Group'>
              group /PRODUCT:
                  dimensions(sizes): scanline(5640), ground_pixel(450), corner(4), time(1), polynomial_exponents(6), layer(34), vertices(2)
                  variables(dimensions): int32 scanline(scanline), int32 ground_pixel(ground_pixel), int32 time(time), int32 corner(corner), int32 polynomial_exponents(polynomial_exponents), int32 layer(layer), int32 vertices(vertices), float32 latitude(time,scanline,ground_pixel), float32 longitude(time,scanline,ground_pixel), int32 delta_time(time,scanline), <class 'str'> time_utc(time,scanline), uint8 qa_value(time,scanline,ground_pixel), float32 nitrogendioxide_tropospheric_column(time,scanline,ground_pixel), float32 nitrogendioxide_tropospheric_column_precision(time,scanline,ground_pixel), float32 averaging_kernel(time,scanline,ground_pixel,layer), float32 air_mass_factor_troposphere(time,scanline,ground_pixel), float32 air_mass_factor_total(time,scanline,ground_pixel), int32 tm5_tropopause_layer_index(time,scanline,ground_pixel), float32 tm5_constant_a(layer,vertices), float32 tm5_constant_b(layer,vertices)
                  groups: SUPPORT_DATA),
             ('METADATA', <class 'netCDF4._netCDF4.Group'>
              group /METADATA:
                  dimensions(sizes): 
                  variables(dimensions): 
                  groups: QA_STATISTICS, ALGORITHM_SETTINGS, GRANULE_DESCRIPTION, ISO_METADATA, EOP_METADATA, ESA_METADATA)])
```
```
# opening the product with xarray raises an UnboundLocalError exception
no2 = xr.open_dataset(product, group='/PRODUCT')
```

```
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-26-2052b95f2d95> in <module>()
----> 1 no2 = xr.open_dataset(product, group='/PRODUCT')

/Users/stefano/anaconda/lib/python3.6/site-packages/xarray/backends/api.py in open_dataset(filename_or_obj, group, decode_cf, mask_and_scale, decode_times, autoclose, concat_characters, decode_coords, engine, chunks, lock, cache, drop_variables)
    303             lock = _default_lock(filename_or_obj, engine)
    304         with close_on_error(store):
--> 305             return maybe_decode_store(store, lock)
    306     else:
    307         if engine is not None and engine != 'scipy':

/Users/stefano/anaconda/lib/python3.6/site-packages/xarray/backends/api.py in maybe_decode_store(store, lock)
    223             store, mask_and_scale=mask_and_scale, decode_times=decode_times,
    224             concat_characters=concat_characters, decode_coords=decode_coords,
--> 225             drop_variables=drop_variables)
    226 
    227         _protect_dataset_variables_inplace(ds, cache)

/Users/stefano/anaconda/lib/python3.6/site-packages/xarray/conventions.py in decode_cf(obj, concat_characters, mask_and_scale, decode_times, decode_coords, drop_variables)
   1153     vars, attrs, coord_names = decode_cf_variables(
   1154         vars, attrs, concat_characters, mask_and_scale, decode_times,
-> 1155         decode_coords, drop_variables=drop_variables)
   1156     ds = Dataset(vars, attrs=attrs)
   1157     ds = ds.set_coords(coord_names.union(extra_coords).intersection(vars))

/Users/stefano/anaconda/lib/python3.6/site-packages/xarray/conventions.py in decode_cf_variables(variables, attributes, concat_characters, mask_and_scale, decode_times, decode_coords, drop_variables)
   1086             k, v, concat_characters=concat_characters,
   1087             mask_and_scale=mask_and_scale, decode_times=decode_times,
-> 1088             stack_char_dim=stack_char_dim)
   1089         if decode_coords:
   1090             var_attrs = new_vars[k].attrs

/Users/stefano/anaconda/lib/python3.6/site-packages/xarray/conventions.py in decode_cf_variable(name, var, concat_characters, mask_and_scale, decode_times, decode_endianness, stack_char_dim)
    998         if (has_fill or scale_factor is not None or add_offset is not None):
    999             if has_fill and np.array(fill_value).dtype.kind in ['U', 'S', 'O']:
-> 1000                 if string_encoding is not None:
   1001                     raise NotImplementedError(
   1002                         'variable %r has a _FillValue specified, but '

UnboundLocalError: local variable 'string_encoding' referenced before assignment
```
```python
# Opening another group with xarray works fine
no2 = xr.open_dataset(product, group='/METADATA/QA_STATISTICS')
no2
```
```
<xarray.Dataset>
Dimensions:                                                (nitrogendioxide_stratospheric_column_histogram_axis: 100, nitrogendioxide_stratospheric_column_pdf_axis: 400, nitrogendioxide_total_column_histogram_axis: 100, nitrogendioxide_total_column_pdf_axis: 400, nitrogendioxide_tropospheric_column_histogram_axis: 100, nitrogendioxide_tropospheric_column_pdf_axis: 400, vertices: 2)
[...]
```
#### Problem description

An UnboundLocalError exception is raised while trying to open a netCDF file on a specific group ('```/PRODUCT```'). Opening the file within another group works fine.
The same file can be correctly opened with the netCDF4 library. 

#### Expected Output
File correctly loaded.

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Darwin
OS-release: 17.2.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

xarray: 0.10.0
pandas: 0.20.1
numpy: 1.12.1
scipy: 0.19.0
netCDF4: 1.2.4
h5netcdf: 0.5.0
Nio: None
bottleneck: 1.2.1
cyordereddict: None
dask: 0.15.3
matplotlib: 2.1.0
cartopy: 0.15.1
seaborn: 0.8.1
setuptools: 27.2.0
pip: 9.0.1
conda: 4.3.29
pytest: 3.0.7
IPython: 5.3.0
sphinx: 1.5.6

</details>

### PR DESCRIPTION
**Pull request by shoyer**
Fixes #1781

The goal here is to eventually convert everything in xarray.conventions to
using the new coding module, which is more modular and supports dask arrays.
For now, I have switched over datetime, timedelta, unsigned integer, scaling
and mask coding to use new coders. Integrating these into xarray.conventions
lets us harness our existing test suite and delete a lot of redundant code.

Most of the code/tests is simply reorganized. There should be no changes to
public API (to keep this manageable for review). All of the original tests that
are still relevant should still be present, though I have reorganized many of
them into new locations to match the revised code.

 - [x] Tests added (for all bug fixes or enhancements)
 - [x] Tests passed (for all non-documentation changes)
 - [x] Passes ``git diff upstream/master **/*py | flake8 --diff`` (remove if you did not edit any Python files)

### CODE REVIEW

File: xarray/conventions.py [hunk 1]
```diff
@@ -958,73 +460,22 @@ def decode_cf_variable(name, var, concat_characters=True, mask_and_scale=True,
...
-                          .format(name, fill_value),
-                          SerializationWarning, stacklevel=3)
-
-        scale_factor = pop_to(attributes, encoding, 'scale_factor')
-        add_offset = pop_to(attributes, encoding, 'add_offset')
-        has_fill = (fill_value is not None and
-                    not np.any(pd.isnull(fill_value)))
-        if (has_fill or scale_factor is not None or add_offset is not None):
-            if has_fill and np.array(fill_value).dtype.kind in ['U', 'S', 'O']:
-                if string_encoding is not None:
```
Comment (shoyer) [inline xarray/conventions.py]: I removed this check. I think I put this logic in the wrong place: we really should ensure that `_FillValue` is not provided when writing/encoding a variable to disk, not in reading/decoding.

File: xarray/backends/zarr.py [hunk 2]
```diff
@@ -259,13 +260,13 @@ def encode_zarr_variable(var, needs_copy=True, name=None):
...
-                                                                name=name)
-    var, needs_copy = conventions.maybe_encode_fill_value(var, needs_copy,
-                                                          name=name)
+    for coder in [coding.times.CFDatetimeCoder(),
+                  coding.times.CFTimedeltaCoder(),
+                  coding.variables.CFScaleOffsetCoder(),
+                  coding.variables.CFMaskCoder(),
+                  coding.variables.UnsignedCoder()]:
+        var = coder.encode(var, name=name)
+
```
Comment (jhamman) [inline xarray/backends/zarr.py]: @shoyer - what do you think about adding encode/decode methods to the `AbstractWritableDataStore`? It seems each backend handles these slightly differently but this step happens for all backends.

Comment (shoyer) [reply xarray/backends/zarr.py]: Yes, some of these need to get associated with the backend classes in some way. I was waiting to do that until we finish porting all of the stuff in `conventions` into `coding`.

File: xarray/coding/variables.py [hunk 3]
```diff
@@ -188,7 +189,89 @@ def decode(self, variable, name=None):
...
+
+        if 'scale_factor' in attrs or 'add_offset' in attrs:
+            scale_factor = pop_to(attrs, encoding, 'scale_factor', name=name)
+            add_offset = pop_to(attrs, encoding, 'add_offset', name=name)
+            dtype = np.float64
+            transform = partial(_scale_offset_decoding,
+                                scale_factor=scale_factor,
+                                add_offset=add_offset,
+                                dtype=dtype)
+            data = lazy_elemwise_func(data, transform, dtype)
```
Comment (rabernat) [inline xarray/coding/variables.py]: This looks great! When did `lazy_elemwise_func` get added? I missed that somehow. This solves our problems related to multiple types of lazy arrays.

**Comment by jhamman** [comment]
@shoyer - can we merge this soon so I can use it in #1800?

File: xarray/coding/variables.py (hunk 3 above)
Comment (jhamman) [reply xarray/coding/variables.py]: @rabernat - see https://github.com/pydata/xarray/pull/1752

**Comment by shoyer** [comment]
I was waiting for more review but if you think this is good to go we can merge it.

I think this also fixes https://github.com/pydata/xarray/issues/1781 -- let me add a regression test.

**Comment by jhamman** [comment]
I'll give it another review.

File: xarray/coding/times.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+    result = pd.to_timedelta(num_timedeltas, unit=units, box=False)
+    # NaT is returned unboxed with wrong units; this should be fixed in pandas
+    if result.dtype != 'timedelta64[ns]':
+        result = result.astype('timedelta64[ns]')
+    return result.reshape(shape)
+
+
+TIME_UNITS = frozenset(['days', 'hours', 'minutes', 'seconds',
+                        'milliseconds', 'microseconds'])
```
[the remark this answers is not in the document]
Comment (shoyer) [reply xarray/coding/times.py]: done

File: xarray/coding/times.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+
+
+def _infer_time_units_from_diff(unique_timedeltas):
+    for time_unit, delta in [('days', 86400), ('hours', 3600),
+                             ('minutes', 60), ('seconds', 1)]:
```
[the remark this answers is not in the document]
Comment (shoyer) [reply xarray/coding/times.py]: Sort of. I refactored it to use `_NS_PER_TIME_DELTA` but it's not that much cleaner than before.

**Comment by jhamman** [comment]
Thanks @shoyer!

**Pull request merged**